src/Lib.P2P/MultiAddressExtensions.cs

```diff
@@ -0,0 +1,147 @@
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using MultiFormats;
+
+namespace Lib.P2P
+{
+    /// <summary>
+    ///   Extensions to <see cref="MultiAddress"/>.
+    /// </summary>
+    public static class MultiAddressExtensions
+    {
+        private static Dictionary<AddressFamily, string> _supportedDnsAddressFamilies =
+            new Dictionary<AddressFamily, string>();
+
+        private static MultiAddress _http = new MultiAddress("/tcp/80");
+        private static MultiAddress _https = new MultiAddress("/tcp/443");
+
+        static MultiAddressExtensions()
+        {
+            if (Socket.OSSupportsIPv4)
+                _supportedDnsAddressFamilies[AddressFamily.InterNetwork] = "/ip4/";
+            if (Socket.OSSupportsIPv6)
+                _supportedDnsAddressFamilies[AddressFamily.InterNetworkV6] = "/ip6/";
+        }
+
+        /// <summary>
+        ///   Determines if the multiaddress references
+        ///   a loopback address.
+        /// </summary>
+        /// <param name="multiaddress">
+        ///   The mutiaddress to clone.
+        /// </param>
+        /// <returns>
+        ///   <b>true</b> for a loopback (127.0.0.1 or ::1).
+        /// </returns>
+        public static bool IsLoopback(this MultiAddress multiaddress)
+        {
+            return multiaddress.Protocols.Any(p =>
+                p.Name == "ip4" && p.Value == "127.0.0.1" ||
+                p.Name == "ip6" && p.Value == "::1");
+        }
+
+        /// <summary>
+        ///   Get all the addresses for the specified <see cref="MultiAddress"/>.
+        /// </summary>
+        /// <param name="multiaddress">
+        ///   The multiaddress to resolve.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result
+        ///   is a sequence of possible multiaddresses.
+        /// </returns>
+        /// <exception cref="SocketException">
+        ///   The host name cannot be resolved.
+        /// </exception>
+        /// <remarks>
+        ///   When the <see cref="NetworkProtocol.Name"/> starts with "dns", then a DNS
+        ///   lookup is performed to get all the IP addresses for the host name.  "dn4" and "dns6"
+        ///   will filter the result for IPv4 and IPV6 addresses.
+        ///   <para>
+        ///   When the <see cref="NetworkProtocol.Name"/> is "http" or "https", then
+        ///   a "tcp/80" or "tcp/443" is respectively added.
+        ///   </para>
+        /// </remarks>
+        public static async Task<List<MultiAddress>> ResolveAsync(this MultiAddress multiaddress,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The Cyclomatic Complexity of this method is 11 which is greater than 10 authorized.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)